monkey/monkey_island/cc/services/config_schema_service.py

```diff
@@ -0,0 +1,79 @@
+from copy import deepcopy
+from typing import Any, Dict, Sequence, Tuple
+
+from common.agent_configuration import AgentConfiguration, PluginConfiguration
+from common.agent_plugins import AgentPlugin, AgentPluginType
+from monkey_island.cc.repositories import IAgentPluginRepository
+
+
+class ConfigSchemaService:
+    """
+    A service for retrieving the agent configuration schema.
+    """
+
+    def __init__(
+        self,
+        agent_plugin_repository: IAgentPluginRepository,
+    ):
+        self._agent_plugin_repository = agent_plugin_repository
+
+    def get_schema(self) -> Dict[str, Any]:
+        """
+        Get the schema.
+
+        :raises RuntimeError: If the schema could not be retrieved
+        """
+        try:
+            agent_config_schema = deepcopy(AgentConfiguration.schema())
+
+            self._add_plugins_to_schema(agent_config_schema)
+
+            return agent_config_schema
+        except Exception as err:
+            raise RuntimeError(err)
+
+    def _add_plugins_to_schema(self, schema: Dict[str, Any]):
+        plugin_catalog = self._agent_plugin_repository.get_plugin_catalog()
+        if plugin_catalog:
+            schema["definitions"].setdefault("AgentPluginsConfiguration", {})
+            schema["definitions"]["AgentPluginsConfiguration"][
+                "title"
+            ] = "AgentPluginsConfiguration"
+            schema["definitions"]["AgentPluginsConfiguration"]["type"] = "object"
+            schema["definitions"]["AgentPluginsConfiguration"]["description"] = (
+                "A configuration for agent plugins.\n It provides a full"
+                " set of available plugins that can be used by the agent.\n"
+            )
+
+        self._set_exploiters_reference(schema, plugin_catalog)
+
+        for (plugin_type, name) in plugin_catalog:
+            plugin = self._agent_plugin_repository.get_plugin(plugin_type, name)
+            plugin_schema = self._create_plugin_schema(plugin)
+
+            plugin_type_name = plugin_type.name.lower()
+            plugin_type_schema = schema["definitions"]["AgentPluginsConfiguration"].setdefault(
+                plugin_type_name, {}
+            )
+            plugin_type_schema.setdefault("anyOf", []).append(plugin_schema)
+
+    def _set_exploiters_reference(
+        self, schema: Dict[str, Any], plugin_catalog: Sequence[Tuple[AgentPluginType, str]]
+    ):
+        plugin_types = set([plugin_type for (plugin_type, _) in plugin_catalog])
+
+        for plugin_type in plugin_types:
+            # Add reference, based on type
+            if plugin_type == AgentPluginType.EXPLOITER:
+                exploitation = schema["definitions"]["ExploitationConfiguration"]
+                brute_force = exploitation["properties"]["brute_force"]
+                brute_force["items"] = {"$ref": "#/definitions/AgentPluginsConfiguration/exploiter"}
```

Comment:
I don't see a reason to add a multi-level path
```suggestion
                brute_force["items"] = {"$ref": "#/exploiters"}
```

Reply:
It is per the specifications: 
<img width="799" alt="image" src="https://user-images.githubusercontent.com/15820737/208619628-6fdaf7ee-b6cc-49e8-86de-050a3ad9ae2f.png">

Reply:
And one line below it says: 
![image](https://user-images.githubusercontent.com/36815064/208646587-96e862a4-50db-46ec-b902-da2c28bc53c7.png)
Also, "as per the spec." references models. Exploiters are not models, they are instances. There can only be one "exploiters" object in the schema, it's not a definition

Reply:
![image](https://user-images.githubusercontent.com/36815064/208650234-0f20afdf-c40e-488f-9b55-5dd5ced6a0ff.png)

Reply:
So, the tl;dr of this is that the spec suggests to put resources in the definitions, but it's not mandatory. We might want to separate those in case we will have actual definitions in the future.
http://json-schema.org/draft/2020-12/json-schema-core.html#name-bundling